framework/src/media/demux/mpeg2ts/TSParser.cpp

```diff
@@ -0,0 +1,654 @@
+	}
+
+	size_t fill = 0;
+	size_t need;
+	while (fill < size) {
+		need = size - fill;
+		if (mPESPacket) {
+			// get remaining payload in last PES packet
+			if (need > mPESParser->ESDataLength() - mPESDataUsed) {
+				need = mPESParser->ESDataLength() - mPESDataUsed;
+			}
+
+			memcpy(&buf[fill], mPESParser->ESData() + mPESDataUsed, need);
+			mPESDataUsed += need;
+			fill += need;
+
+			if (mPESDataUsed == mPESParser->ESDataLength()) {
+				// read whole es data in one pes packet
+				delete mPESPacket;
+				mPESPacket = NULL;
+				mPESDataUsed = 0;
+			}
+			continue;
+		}
+
+		// get new PES packet
+		if (getPESPacket(&mPESPacket) <= 0) {
+			// 0 / -1
+			break;
+		}
+
+		// parse PES packet
+		if (mPESParser->Parse(mPESPacket->Data(), mPESPacket->DataLength())) {
+			mPESDataUsed = 0;
+		} else {
+			meddbg("PES parse failed\n");
+			assert(0);
+		}
+	}
+
+	return fill;
+}
+
+bool TSParser::getPrograms(std::vector<prog_num_t> &progs)
+{
+	return mParserManager->GetPrograms(progs);
+}
+
+audio_type_t TSParser::getAudioType(prog_num_t progNum)
+{
+	uint8_t streamType;
+	ts_pid_t streamPid;
+	if (mParserManager->GetAudioStreamInfo(progNum, streamType, streamPid)) {
+		switch (streamType) {
+		case PMTElementary::STREAM_TYPE_AUDIO_AAC:
+		case PMTElementary::STREAM_TYPE_AUDIO_HE_AAC:
+			return AUDIO_TYPE_AAC;
+		case PMTElementary::STREAM_TYPE_AUDIO_MPEG1:
+			return AUDIO_TYPE_MP3;
+		default:
+			meddbg("unsupported audio type 0x%x\n", streamType);
+			break;
+		}
+	} else {
+		meddbg("didn't get any audio stream info\n");
+	}
+
+	return AUDIO_TYPE_INVALID;
+}
+
+int TSParser::Adjust(uint8_t *pPacketData, size_t readOffset)
+{
+	uint8_t buffer[TS_PACKET_SIZE];
+	TSHeader tsHeader;
+	size_t szRead;
+	int syncOffset;
+	int count;
+
+	for (syncOffset = 0; syncOffset < TS_PACKET_SIZE; syncOffset++) {
+		if (pPacketData[syncOffset] != SYNCCODE) {
+			continue;
+		}
+
+		// found sync byte, now do sync verification!
+		for (count = 1; count < SYNC_COUNT; count++) {
+			szRead = mBufferReader->copy(buffer, TS_PACKET_SIZE, readOffset + syncOffset + count * TS_PACKET_SIZE);
+			if (szRead != TS_PACKET_SIZE) {
+				// data in buffer is not enough for sync verification
+				return -1;
+			}
+
+			if (!tsHeader.Parse(buffer)) {
+				// sync not match
+				break;
+			}
+		}
+
+		if (count == SYNC_COUNT) {
+			// sync verification succeed
+			return syncOffset;
+		}
+	}
+
+	// Do not find sync byte in one ts packet (188 bytes),
+	// maybe it's not a transport stream.
+	return TS_PACKET_SIZE;
+}
+
+bool TSParser::PSIUnpack(TSHeader &tsHeader, Section **ppSection)
+{
+	Section *pSection = NULL;
+	std::map<int, Section *>::iterator it;
+
+	AdaptationField adaptation_field;
+	uint8_t *pu8Payload = &m_data[4];
+	uint8_t  lenPayload = 188 - 4;
+
+	assert(ppSection != nullptr);
+
+	do {
+		if (tsHeader.TransportErrorIndicator() != 0) {
+			// error
+			break;
+		}
+
+		if (tsHeader.AdaptationFieldControl() == 0) {
```

Comment:
I think 0, 1, 2, 3 is common number for filed control why don't you use enum instead of constant?